File: contest/LeetCode_Weekly_Contest_204/Minimum_Number_of_Days_to_Disconnect_Island.py
class Solution(object):
    dr = [0, -1, 0, 1]
    dc = [-1, 0, 1, 0]
    
    def dfs(self, r, c):
        self.vis[r][c] = 1
        self.cnt += 1
        
        for k in xrange(4):
            x = r + self.dr[k]
            y = c + self.dc[k]
            if 0 <= x and x < self.n and 0 <= y and y < self.m and self.grid[x][y] and not self.vis[x][y]:
                self.dfs(x, y)
        
    def check(self, tot):
        self.vis = [[0 for _ in xrange(self.m)] for _ in xrange(self.n)]
        self.cnt = 0
        for i in xrange(0, self.n):
            for j in xrange(0, self.m):
                if self.grid[i][j]:
                    self.dfs(i, j)
                    if self.cnt != tot:
                        return True
                    else:
                        return False
        return False
    
    def minDays(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        
        self.n = len(grid)
        self.m = len(grid[0])
        self.grid = grid
        
        onecnt = 0
        for i in xrange(0, self.n):
            for j in xrange(0, self.m):
                if self.grid[i][j]: 
                    onecnt += 1
        
        if self.check(onecnt):
            return 0
        
        if onecnt == 1:
            return 1
        
        for i in xrange(0, self.n):
            for j in xrange(0, self.m):
                if self.grid[i][j]:
                    self.grid[i][j] = 0
                    if self.check(onecnt - 1):
                        return 1
                    self.grid[i][j] = 1
        return 2

File: contest/LeetCode_Weekly_Contest_204/Minimum_Number_of_Days_to_Disconnect_Island.js
/**
 * @param {number[][]} grid
 * @return {number}
 */

const dr = [0, -1, 0, 1];
const dc = [-1, 0, 1, 0];

var vis = null
var cnt = null

function makeArr(n, m) {
    var a = new Array()
    for (var i = 0; i < n; ++i) {
        a[i] = new Array()
        for (var j = 0; j < m; ++j)
            a[i][j] = 0
    }
    return a
}

function dfs(grid, r, c) {
    vis[r][c] = true;
    cnt++;
    for (var k = 0; k < 4; ++k) {
        var x = r + dr[k];
        var y = c + dc[k];
        if (0 <= x && x < n && 0 <= y && y < m && grid[x][y] && !vis[x][y])
            dfs(grid, x, y);
    }
}

function check(grid, tot) {
    for (var i = 0; i < n; ++i)
        for (var j = 0; j < m; ++j) if (grid[i][j]) {
            cnt = 0;
            dfs(grid, i, j);
            if (cnt !== tot) return true;
            else return false;
        }
    return false;
}

var minDays = function(grid) {
    n = grid.length;
    m = grid[0].length;
        
    onecnt = 0;
    for (var i = 0; i < n; ++i)
        for (var j = 0; j < m; ++j)
            if (grid[i][j]) onecnt++;
        
    if (onecnt == 1) return 1;
        
    vis = makeArr(n, m)
    if (check(grid, onecnt)) return 0;
        
    for (var i = 0; i < n; ++i)
        for (var j = 0; j < m; ++j) if (grid[i][j]) {
            vis = makeArr(n, m)
            grid[i][j] = 0;
            if (check(grid, onecnt-1)) return 1;
            grid[i][j] = 1;
        }
        
    return 2;
};